File: os/Sem.java
import java.util.concurrent.Semaphore;

public class Sem {
    static int sharedData = 0; // 공유 데이터
    // 세마포 생성, 공유 자원 1개
    static Semaphore semaphore = new Semaphore(1);

    public static void main(String[] args) {
        Thread thread1 = new Thread(new Increment());
        Thread thread2 = new Thread(new Decrement());

        thread1.start(); // 첫 번째 스레드 시작
        thread2.start(); // 두 번째 스레드 시작
        try {
            thread1.join(); // 첫 번째 스레드 종료 대기
            thread2.join(); // 두 번째 스레드 종료 대기
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("Final value of sharedData: " + sharedData);
    }

    static class Increment implements Runnable {
        public void run() {
            for (int i = 0; i < 100000; i++) {
                try {
                    semaphore.acquire(); // 세마포 획득
                    sharedData++; // 공유 데이터 증가
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } finally {
                    semaphore.release(); // 세마포 해제
                }
            }
        }
    }

    static class Decrement implements Runnable {
        public void run() {
            for (int i = 0; i < 100000; i++) {
                try {
                    semaphore.acquire(); // 세마포 획득
                    sharedData--; // 공유 데이터 감소
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } finally {
                    semaphore.release(); // 세마포 해제
                }
            }
        }
    }
}

File: os/sem.cpp
#include <semaphore.h>
#include <iostream>
#include <thread>

int sharedData = 0;  // 공유 데이터
sem_t semaphore;     // 세마포어 선언

void increment() {
    for (int i = 0; i < 100000; i++) {
        sem_wait(&semaphore);  // 세마포어 획득
        sharedData++;          // 공유 데이터 증가
        sem_post(&semaphore);  // 세마포어 해제
    }
}

void decrement() {
    for (int i = 0; i < 100000; i++) {
        sem_wait(&semaphore);  // 세마포어 획득
        sharedData--;          // 공유 데이터 감소
        sem_post(&semaphore);  // 세마포어 해제
    }
}

int main() {
    sem_init(&semaphore, 0, 1);  // 세마포어 초기화, 공유 자원 1개
    std::thread thread1(increment);
    std::thread thread2(decrement);
    thread1.join();
    thread2.join();
    std::cout << "Final value of sharedData: " << sharedData << std::endl;

    // 세마포어 제거
    sem_destroy(&semaphore);
    return 0;
}
